### ISSUE TITLE
given() callback automatically executed for each test but the first

### ISSUE THREAD

**Issue by LostKobrakai** [kahlan/kahlan#207]
I've toyed around with the given() functionality and noticed, that the callbacks are executed for each test before it actually does start. I think this has something to do with how the setup of those is handled, as it's not happening for the first executed test.

The test/output cen be found here: https://github.com/LostKobrakai/kahlan-issue

**Comment by jails** [comment]
`given()` is a direct port of RSpec's `let` and the its main characteristic is that variables are lazily loaded and also cached across a same scope (but not across specifications where they are lazily when necessary).

Let's take the following example:

```php
given('variable', function () {
    return [];
});
given('variable', function () {
    $this->variable[] = 1;
    return $this->variable;
});
given('variable', function () {
    $this->variable[] = 2;
    return $this->variable;
});
expect($this->variable)->toBe([1, 2]);
```
The above example shows that `variable` is lazy loaded recursively inside the same scope.

So with your example:

```php
beforeEach(function () {
    given('laravel', function() {
        var_dump('Init laravel   ');
	return (object) ['app' => 'Test'];
    });
});
```
The first time is fine but on the second spec the `beforeEach()` is executed once again and the fact the `laravel` variable already exists in this scope Kahlan simply rebuilds it (and load its previous value just in case you are going to use it inside your new `given()` definition).

Long story short you shouldn't mix `beforeEach()` and `given()` all together. `given()` are definitions and don't need to be redefined. Moving `given()` definitions apart should make your specs to work as expected:

```php
given('laravel', function() {
    ...
});
given('app', function() {
    ...
});
beforeEach(function () {
    ...
});
```

**Comment by LostKobrakai** [comment]
I've no experience with rspec to draw from, but that does make sense. Putting those given() calls outside did indeed resolve those issues. Thanks for the rundown. I'll later try to create a pr making this more clear in the docs.

**Comment by jails** [comment]
You bet!

**Comment by LostKobrakai** [comment]
I've another issue/request regarding how `given()` does work. Would it be possible if a `given()` variable is used/resolved as part of a `beforeEach()` callback, that it's still only kept for a single "spec" (a.k.a. `beforeEach()`/`it()`/`afterEach()` round) and not for the whole `describe()` block?

```php
given('processwire', function() { … });

// Partly pseudo code
describe('some feature', function() {
	describe('using migrations', function() {
		beforeEach(function() {
			$this->processwire->run->migrations;
		});
		it('should migrate for each test', function() {
			expect($this->processwire)->toBeAn('object');
		});
		it('should migrate for each test', function() {
			expect($this->processwire)->toBeANew('object');
		});
	});
});
```

I can see that a beforeEach callback is more related to the describe block, but I'd like to abstract the beforeEach() callback away into a library, so it would be really nice to be able to scope those calls to a single spec, while not starting the processwire instance for each and every test.

When using in a `beforeAll()` it should be clear enough that things are resolved once for "all" tests associated to that `beforeAll()`.

**Comment by jails** [comment]
Good catch :+1: 
Should be fixed in https://github.com/kahlan/kahlan/commit/73ef4bae73981a01a7661c17a7ab61c90930bfd0

**Comment by jails** [comment]
Legit issue :+1: 
Should be fixed in https://github.com/kahlan/kahlan/commit/e961bba9247326102afcb50e12d206e1124c83ab

**Comment by LostKobrakai** [comment]
Now it's doing what I though it would do (correctly) above, but I'd actually see it useful to have `beforeEach()` / `it()` and `afterEach()` to receive the same instance instead of a new one each time. Say I have a global `given('processwire', function () {});` to bootstrap the app. In some tests I might use `$this->processwire` without any `beforeEach()` using that instance, but on others I might need to run those migrations on that instance (in the `beforeEach()`) before running the tests. 

Currently I could save the migrated instance under a new name in the beforeEach like so, but then I'd need to access the instance differently in cases with a migrated context than in cases without. Would be nicer to just use the same name, especially if the beforeEach is declared somewhere in an installable package and not directly in the test.
```php
given('processwire', function() { … });

describe('some feature', function() {
	it('access unmigrated instance', function() {
		expect($this->processwire)->toBeAn('object');
	});
	describe('using migrations', function() {
		beforeEach(function() {
			$app = $this->processwire;
			// Migrate stuff
			$this->migrated = $app;
		});
		it('should migrate this instance before the test', function() {
			expect($this->migrated)->toBeAn('object');
		});
		it('should migrate another instance before the test', function() {
			expect($this->migrated)->toBeANew('object');
		});
	});
});
```

I'm not sure how in line this would be with the overall design decisions to alter/add such a behaviour. Just throwing my thoughts on that out here.

**Comment by jails** [comment]
A yeah good catch, `beforeEach()`, `it()` and `afterEach()` should be considered as a unique scope. Should be fixed in https://github.com/kahlan/kahlan/commit/45ebdf533205b9bf8ff7d2ae1f2d0f2261060247

**Comment by LostKobrakai** [comment]
I've expanded on this in #208, so it does clear values of _given setups higher up in parent scopes.
